src/common/code_utils.hpp

```diff
@@ -175,4 +175,62 @@ class NonCopyable
     NonCopyable(void) = default;
 };
 
+template <typename T> class Optional
+{
+public:
+    constexpr Optional(void) = default;
+
+    Optional(T aValue)
+    {
+        new (&mStorage) T(aValue);
+        mHasValue = true;
+    }
+
+    ~Optional(void)
+    {
+        if (mHasValue)
+        {
+            GetValue().~T();
+        }
+    }
+
+    Optional(const Optional &aOther)
+    {
+        if (aOther.mHasValue)
+        {
+            new (&mStorage) T(aOther.GetValue());
+        }
+        mHasValue = aOther.mHasValue;
+    }
+
+    Optional &operator=(const Optional &aOther)
+    {
+        if (mHasValue)
+        {
+            GetValue().~T();
+        }
+        if (aOther.mHasValue)
+        {
+            new (&mStorage) T(aOther.GetValue());
+        }
+        mHasValue = aOther.mHasValue;
+    }
+
+    constexpr const T *operator->(void)const { return &GetValue(); }
+
+    constexpr const T &operator*(void)const { return GetValue(); }
+
+    constexpr bool HasValue(void) const { return mHasValue; }
+
+private:
+    T &GetValue(void) const
+    {
+        VerifyOrDie(mHasValue, "Optional doesn't contain a value");
```

Comment:
why not use assert?
I think alling `GetValue` when there is no value is an logical error, thus `assert` should be most suitable.

Reply:
`assert` depends on the build configuration. I prefer to make it a consistent error checking, so users of this class won't trigger undefined behavior no matter how they use this class. Thoughts?

Reply:
I don't think that's a reason to not use `assert`. 

`assert` should be used for logical errors, while `VerifyOrDie` should be used for fatal errors that can't be recovered.

Reply:
> I think it's a good practice to never turn off assert in release builds, but that might be another topic.

If you have this idea in your mind, it means you are wanting something which shouldn't be provided by `assert`. `assert` should be just treated as a comment for helping understand the logic (i.e. what assumptions this method takes to work properly). When deciding whether we should use `assert` or `VerifyOrDie`, ask yourself if simply leaving a comment here is what you want.

Reply:
I think we are already using `OT_ASSERT` extensively in OpenThread to catch logical errors. 
The default definition of `OT_ASSERT` on Posix platform is `assert`, but it's possible to provide platform-specific assert behaviors if the default `assert` is not exactly what you want.
 
So, I think it's the same practice in ot-br-posix that we should use `assert`. 

Maybe we should also define `OTBR_ASSERT` to replace the standard library `assert`, but I guess it's beyond the scope of this PR.

Reply:
I agree with @wgtdkp that assertions should not be relied on, because in release builds they are indeed just comments.

For this particular case, I looked up the code base and found that `VerifyOrDie` is for unrecoverable external errors. I think it's also OK to treat this as a comment. Although that would require developers to be more cautious, being consistent on the function purposes is probably more important. Updated.